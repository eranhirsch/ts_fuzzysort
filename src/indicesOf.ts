/**
 * Similar to `String.prototype.indexOf`, but performs the search in sequence
 * for each code point in the needles. This means that returned indices are in
 * order.
 *
 * @param needles The sequence to being searched for.
 * @param haystack The sequence to search in.
 * @returns A sequence of indices that represent a projection of needles on
 * haystack; or `undefined` if no such sequence exists.
 */
export function indicesOf<T>(
  needles: readonly T[],
  haystack: readonly T[]
): readonly number[] | undefined {
  const output: number[] = [];

  for (const [searchableIndex, searchableCodePoint] of haystack.entries()) {
    if (needles[output.length] !== searchableCodePoint) {
      continue;
    }

    output.push(searchableIndex);
    if (output.length === needles.length) {
      // We've found a match for all code points in the query
      return output;
    }
  }

  // We haven't found a match for all code points in the query in the same
  // order, so this searchable doesn't contain a match (even fuzzy!)
  return undefined;
}
