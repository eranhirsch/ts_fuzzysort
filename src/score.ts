/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable security/detect-object-injection */

import { Query, Searchable } from "./text";

const PENALTY_MATCH_GROUPS = 12;
const PENALTY_FIRST_MATCHING_OFFSET = 0.2;

const PENALTY_NON_STRICT_MATCH = 1000;

const LONG_PHRASES_THRESHOLD = 24;
const LONG_PHRASES_PENALTY = 10;

export function matchScore(
  query: Omit<Query, "presentCharacters">,
  searchable: Searchable,
  match: readonly number[],
  isStrict: boolean,
  nextWordBreak: readonly number[]
): number {
  let score = 0;

  score -= multiGroupPenalty(match, query.codePoints.length);

  score -= offsetPenalty(match);

  score *= isStrict ? phrasesPenalty(nextWordBreak) : PENALTY_NON_STRICT_MATCH;

  score /= substringBonus(query, searchable, match, nextWordBreak);

  // penalty for longer targets
  score -= searchable.codePoints.length - query.codePoints.length;

  return score;
}

// penalty for more groups
function multiGroupPenalty(
  match: readonly number[],
  queryLength: number
): number {
  let groupPositionPenalty = 0;

  let extraMatchGroupCount = 0;
  for (let index = 1; index < match.length; index += 1) {
    if (match[index]! !== match[index - 1]! + 1) {
      groupPositionPenalty += match[index]!;
      ++extraMatchGroupCount;
    }
  }

  const unmatchedDistance =
    match[queryLength - 1]! - match[0]! - (queryLength - 1);

  const distancePenalty =
    (PENALTY_MATCH_GROUPS + unmatchedDistance) * extraMatchGroupCount;

  return groupPositionPenalty + distancePenalty;
}

// penalty for not starting near the beginning
function offsetPenalty(match: readonly number[]): number {
  const firstMatchingIndex = match[0]!;

  if (firstMatchingIndex === 0) {
    return 0;
  }

  return (
    firstMatchingIndex * firstMatchingIndex * PENALTY_FIRST_MATCHING_OFFSET
  );
}

// successStrict on a target with too many beginning indexes loses points
// for being a bad target
function phrasesPenalty(nextWordBreak: readonly number[]): number {
  let words = 1;
  for (
    let index = nextWordBreak[0]!;
    index < nextWordBreak.length;
    index = nextWordBreak[index]!
  ) {
    words += 1;
  }

  if (words <= LONG_PHRASES_THRESHOLD) {
    return 1;
  }

  // quite arbitrary numbers here ...
  return (words - LONG_PHRASES_THRESHOLD) * LONG_PHRASES_PENALTY;
}

function substringBonus(
  query: Omit<Query, "presentCharacters">,
  searchable: Searchable,
  match: readonly number[],
  nextWordBreak: readonly number[]
): number {
  const firstMatchingIndex = match[0]!;
  if (!searchable.lowerCase.startsWith(query.lowerCase, firstMatchingIndex)) {
    return 1;
  }

  // bonus for being a full substring
  const bonus = 1 + query.codePoints.length * query.codePoints.length * 1;

  if (
    firstMatchingIndex === 0 ||
    nextWordBreak[firstMatchingIndex - 1] === firstMatchingIndex
  ) {
    return bonus;
  }
  // bonus for substring starting on a beginningIndex

  return bonus * (1 + query.codePoints.length * query.codePoints.length * 1);
}
