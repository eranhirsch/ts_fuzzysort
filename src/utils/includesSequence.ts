/* eslint-disable security/detect-object-injection */

export function includesSequence<T>(
  data: readonly T[],
  sequence: readonly T[],
  startIndex = 0
): boolean {
  let sequenceIndex = 0;

  for (
    let index = startIndex;
    index < data.length - (sequence.length - sequenceIndex - 1);
    index += 1
  ) {
    if (data[index] === sequence[sequenceIndex]) {
      if (sequenceIndex === sequence.length - 1) {
        return true;
      }

      sequenceIndex += 1;
    } else {
      sequenceIndex = 0;
    }
  }

  return false;
}
