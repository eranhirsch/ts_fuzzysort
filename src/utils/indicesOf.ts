/**
 * Similar to `String.prototype.indexOf`, but performs the search in sequence
 * for each code point in the needles. This means that returned indices are in
 * order.
 *
 * @param needles The sequence to being searched for.
 * @param haystack The sequence to search in.
 * @returns A sequence of indices that represent a projection of needles on
 * haystack; or `undefined` if no such sequence exists.
 */
export function indicesOf<Needles extends readonly unknown[] | []>(
  needles: Needles,
  haystack: readonly Needles[number][]
): { readonly [P in keyof Needles]: number } | undefined {
  const output: number[] = [];

  for (const [searchableIndex, searchableCodePoint] of haystack.entries()) {
    if (needles[output.length] !== searchableCodePoint) {
      continue;
    }

    output.push(searchableIndex);
    if (output.length === needles.length) {
      // We've found a match for all code points in the query
      // @ts-expect-error [ts2322] - We build output incrementally so typescript can't ensure that it matches the expected output.
      return output;
    }
  }

  // We haven't found a match for all code points in the query in the same
  // order, so this searchable doesn't contain a match (even fuzzy!)
  return undefined;
}
