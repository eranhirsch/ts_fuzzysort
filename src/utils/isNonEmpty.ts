export type NonEmptyArray<T> = readonly [T, ...T[]];

export const isNonEmpty = <T>(items: readonly T[]): items is NonEmptyArray<T> =>
  items.length > 0;
