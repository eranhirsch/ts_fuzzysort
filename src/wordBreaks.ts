export function nextWordBreakIndices(input: string): readonly number[] {
  const characters = [...input];

  const output: number[] = Array.from({ length: characters.length });

  let previousWordBreakIndex = 0;
  for (const wordBreakIndex of findWordBreakIndices(characters)) {
    output.fill(wordBreakIndex, previousWordBreakIndex, wordBreakIndex);
    previousWordBreakIndex = wordBreakIndex;
  }
  output.fill(characters.length, previousWordBreakIndex);

  return output;
}

export function* findWordBreakIndices(
  characters: readonly string[],
): Generator<number, undefined> {
  let previousWasUpperCase = false;
  let previousAlphaNumeric = false;

  for (const [index, character] of characters.entries()) {
    // Would these be more efficient as a regex? Does JS offer any helpers?
    const isUpper = character >= "A" && character <= "Z";
    const isAlphaNumeric =
      isUpper ||
      (character >= "a" && character <= "z") ||
      (character >= "0" && character <= "9");

    if (
      index > 0 &&
      ((isUpper && !previousWasUpperCase) ||
        !previousAlphaNumeric ||
        !isAlphaNumeric)
    ) {
      yield index;
    }

    previousWasUpperCase = isUpper;
    previousAlphaNumeric = isAlphaNumeric;
  }

  return undefined;
}
