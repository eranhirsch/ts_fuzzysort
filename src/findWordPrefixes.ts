/* eslint-disable security/detect-object-injection */

import { type NonEmptyArray } from "./utils/isNonEmpty";

const MAX_BACKTRACK_ATTEMPTS = 200;

/**
 * Look for the query as a set of word prefixes in searchable.
 *
 * This function takes nextWordBreaks and firstMatchingIndex as arguments
 * instead of computing them itself for performance reasons. They are already
 * present at the caller so we don't need to run the logic twice.
 *
 * @param queryCharacters Characters of the lowercased query.
 * @param textCharacters Characters of the lowercased searchable.
 * @param nextWordBreak [Precomputed] An array of indices to the next word, for
 * each given index of searchable.
 * @param firstMatchingIndex The location of the first code point in query
 * inside searchable (the equivalent of running `searchable.indexOf(query[0])`)
 * @returns A sequence of indices that represent a projection of query on
 * searchable; or `undefined` if no such sequence exists.
 */
export function findWordPrefixes(
  // We take characters instead of a raw string because String.prototype.length
  // doesn't take multi-code point characters into account.
  queryCharacters: NonEmptyArray<string>,
  textCharacters: readonly string[],
  nextWordBreak: readonly number[],
  firstMatchingIndex: number,
): NonEmptyArray<number> | undefined {
  let index =
    firstMatchingIndex === 0 ? 0 : nextWordBreak[firstMatchingIndex - 1];

  if (index === undefined || index >= textCharacters.length) {
    // The whole match is already part of the last (or only) word.
    return;
  }

  // We use the length of output as an index into the query, so we can't create
  // a sparse array here.
  const output: number[] = [];
  let backtrackAttempts = 0;

  for (;;) {
    if (
      index !== undefined &&
      queryCharacters[output.length] === textCharacters[index]
    ) {
      // We have match within the current word
      output.push(index);
      if (output.length === queryCharacters.length) {
        // @ts-expect-error [ts2322] - Output is guaranteed to have at least 1 element, we just pushed it!
        return output;
      }

      index += 1;
      continue;
    }

    index = index === undefined ? undefined : nextWordBreak[index];
    if (index !== undefined && index < textCharacters.length) {
      // We found a mismatch in the current word so we skip to the next word
      continue;
    }

    // We've scanned the whole searchable and weren't able to match word
    // word prefixes. We can try "skipping earlier" to see if we can match
    // prefixes differently.

    backtrackAttempts += 1;
    if (backtrackAttempts > MAX_BACKTRACK_ATTEMPTS) {
      // We've attempted to many variations, let's give up so we remain quick.
      return;
    }

    // We pop the last match from the result so we can try matching it with a
    // later word in searchable.
    const lastMatch = output.pop();
    if (lastMatch === undefined) {
      // We've retried all possible variations, the query simply doesn't exist
      // as a set of word prefixes in searchable.
      return;
    }

    index = nextWordBreak[lastMatch];
  }
}
