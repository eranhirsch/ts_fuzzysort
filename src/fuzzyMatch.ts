/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { findWordPrefixes } from "./findWordPrefixes";
import { matchScore } from "./score";
import { indicesOf } from "./utils/indicesOf";
import { type NonEmptyArray } from "./utils/isNonEmpty";
import { nextWordBreakIndices } from "./wordBreaks";

export interface FuzzyMatch {
  readonly score: number;
  readonly indices: readonly number[];
}

export function fuzzyMatch(
  query: NonEmptyArray<string>,
  textRaw: string
): FuzzyMatch | undefined {
  const text = [...textRaw.toLowerCase()];
  const matchSequence = indicesOf(query, text);
  if (matchSequence === undefined) {
    return;
  }

  const nextWordBreak = nextWordBreakIndices(textRaw);

  const wordPrefixesMatchSequence = findWordPrefixes(
    query,
    text,
    nextWordBreak,
    matchSequence[0]!
  );

  const indices = wordPrefixesMatchSequence ?? matchSequence;

  const score = matchScore(
    query,
    text,
    indices,
    wordPrefixesMatchSequence !== undefined,
    nextWordBreak
  );

  return { score, indices };
}
