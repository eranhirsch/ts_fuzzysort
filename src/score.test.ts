/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { asCharactersArray } from "../test/asCharactersArray";
import { findWordPrefixes } from "./findWordPrefixes";
import { fuzzyMatch } from "./fuzzyMatch";
import { matchScore } from "./score";
import { indicesOf } from "./utils/indicesOf";
import { type NonEmptyArray } from "./utils/isNonEmpty";
import { nextWordBreakIndices } from "./wordBreaks";

describe("legacy", () => {
  test("exact match", () => {
    const input = "this is exactly the same search and target";
    expect(simpleMatchScore(input, input)).toBe(0);
  });

  test("scoring", () => {
    const query = "note";

    const score1 = simpleMatchScore(query, "node/NoTe");
    const score2 = simpleMatchScore(query, "not one that evening");

    expect(score1).greaterThan(score2);
  });

  test("substring not at start of word", () => {
    const query = "er.life360";

    const score1 = simpleMatchScore(query, "device-tracker.life360_iphone_6");
    const score2 = simpleMatchScore(query, "sendor.battery_life360_iphone_6");

    expect(score1).greaterThan(score2);
  });

  test("Partial queries yield lower scores", () => {
    const input = "The Amazing Spider-Man";

    const theAmazingSpiderScore = simpleMatchScore("The Amazing Spider", input);
    expect(theAmazingSpiderScore).toBeLessThan(0 /* score for exact match */);

    const theAmazingScore = simpleMatchScore("The Amazing", input);
    expect(theAmazingSpiderScore).toBeGreaterThan(theAmazingScore);

    const theScore = simpleMatchScore("The", input);
    expect(theAmazingScore).toBeGreaterThan(theScore);
  });
});

const simpleMatchScore = (query: string, target: string) =>
  matchScore(
    ...validatedParametersForMatchScore(asCharactersArray(query), target),
  );

/**
 * It's really hard to test the score method because it relies on so many of the
 * computations performed in the fuzzyMatch implementation. The only way we can
 * isolate it is by running the same logic in the test (by re-implementing it)
 * and then asserting it matches saved snapshots with as many of the results as
 * possible. Then we also need to assert it matches the result of running the
 * actual fuzzyMatch logic so that we can assert that the 2 implementations
 * haven't diverged.
 */
function validatedParametersForMatchScore(
  query: NonEmptyArray<string>,
  text: string,
) {
  const parameters = fakeFuzzyMatchResults(query, text);
  expect(parameters).toMatchSnapshot();
  const { indices } = fuzzyMatch(query, text) ?? {};
  expect(indices).toMatchSnapshot();

  expect(parameters).toBeDefined();

  return parameters!;
}

/**
 * This function is an almost exact clone of `fuzzyMatch` until the point it
 * calls `matchScore`, at that point it returns the parameters it would have
 * sent to matchScore so that we can call the function ourselves as part of our
 * tests.
 */
function fakeFuzzyMatchResults(query: NonEmptyArray<string>, textRaw: string) {
  const text = [...textRaw.toLowerCase()];
  const matchSequence = indicesOf(query, text);
  if (matchSequence === undefined) {
    return;
  }

  const nextWordBreak = nextWordBreakIndices(textRaw);

  const wordPrefixesMatchSequence = findWordPrefixes(
    query,
    text,
    nextWordBreak,
    matchSequence[0]!,
  );

  const indices = wordPrefixesMatchSequence ?? matchSequence;

  return [
    query,
    text,
    indices,
    wordPrefixesMatchSequence !== undefined,
    nextWordBreak,
  ] as const;
}
