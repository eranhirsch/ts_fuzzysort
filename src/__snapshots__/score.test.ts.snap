// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`legacy > Partial queries yield lower scores 1`] = `
[
  [
    "t",
    "h",
    "e",
    " ",
    "a",
    "m",
    "a",
    "z",
    "i",
    "n",
    "g",
    " ",
    "s",
    "p",
    "i",
    "d",
    "e",
    "r",
  ],
  [
    "t",
    "h",
    "e",
    " ",
    "a",
    "m",
    "a",
    "z",
    "i",
    "n",
    "g",
    " ",
    "s",
    "p",
    "i",
    "d",
    "e",
    "r",
    "-",
    "m",
    "a",
    "n",
  ],
  [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
  ],
  true,
  [
    3,
    3,
    3,
    4,
    11,
    11,
    11,
    11,
    11,
    11,
    11,
    12,
    18,
    18,
    18,
    18,
    18,
    18,
    19,
    22,
    22,
    22,
  ],
]
`;

exports[`legacy > Partial queries yield lower scores 2`] = `
[
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
]
`;

exports[`legacy > Partial queries yield lower scores 3`] = `
[
  [
    "t",
    "h",
    "e",
    " ",
    "a",
    "m",
    "a",
    "z",
    "i",
    "n",
    "g",
  ],
  [
    "t",
    "h",
    "e",
    " ",
    "a",
    "m",
    "a",
    "z",
    "i",
    "n",
    "g",
    " ",
    "s",
    "p",
    "i",
    "d",
    "e",
    "r",
    "-",
    "m",
    "a",
    "n",
  ],
  [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
  ],
  true,
  [
    3,
    3,
    3,
    4,
    11,
    11,
    11,
    11,
    11,
    11,
    11,
    12,
    18,
    18,
    18,
    18,
    18,
    18,
    19,
    22,
    22,
    22,
  ],
]
`;

exports[`legacy > Partial queries yield lower scores 4`] = `
[
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
]
`;

exports[`legacy > Partial queries yield lower scores 5`] = `
[
  [
    "t",
    "h",
    "e",
  ],
  [
    "t",
    "h",
    "e",
    " ",
    "a",
    "m",
    "a",
    "z",
    "i",
    "n",
    "g",
    " ",
    "s",
    "p",
    "i",
    "d",
    "e",
    "r",
    "-",
    "m",
    "a",
    "n",
  ],
  [
    0,
    1,
    2,
  ],
  true,
  [
    3,
    3,
    3,
    4,
    11,
    11,
    11,
    11,
    11,
    11,
    11,
    12,
    18,
    18,
    18,
    18,
    18,
    18,
    19,
    22,
    22,
    22,
  ],
]
`;

exports[`legacy > Partial queries yield lower scores 6`] = `
[
  0,
  1,
  2,
]
`;

exports[`legacy > exact match 1`] = `
[
  [
    "t",
    "h",
    "i",
    "s",
    " ",
    "i",
    "s",
    " ",
    "e",
    "x",
    "a",
    "c",
    "t",
    "l",
    "y",
    " ",
    "t",
    "h",
    "e",
    " ",
    "s",
    "a",
    "m",
    "e",
    " ",
    "s",
    "e",
    "a",
    "r",
    "c",
    "h",
    " ",
    "a",
    "n",
    "d",
    " ",
    "t",
    "a",
    "r",
    "g",
    "e",
    "t",
  ],
  [
    "t",
    "h",
    "i",
    "s",
    " ",
    "i",
    "s",
    " ",
    "e",
    "x",
    "a",
    "c",
    "t",
    "l",
    "y",
    " ",
    "t",
    "h",
    "e",
    " ",
    "s",
    "a",
    "m",
    "e",
    " ",
    "s",
    "e",
    "a",
    "r",
    "c",
    "h",
    " ",
    "a",
    "n",
    "d",
    " ",
    "t",
    "a",
    "r",
    "g",
    "e",
    "t",
  ],
  [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30,
    31,
    32,
    33,
    34,
    35,
    36,
    37,
    38,
    39,
    40,
    41,
  ],
  true,
  [
    4,
    4,
    4,
    4,
    5,
    7,
    7,
    8,
    15,
    15,
    15,
    15,
    15,
    15,
    15,
    16,
    19,
    19,
    19,
    20,
    24,
    24,
    24,
    24,
    25,
    31,
    31,
    31,
    31,
    31,
    31,
    32,
    35,
    35,
    35,
    36,
    42,
    42,
    42,
    42,
    42,
    42,
  ],
]
`;

exports[`legacy > exact match 2`] = `
[
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  21,
  22,
  23,
  24,
  25,
  26,
  27,
  28,
  29,
  30,
  31,
  32,
  33,
  34,
  35,
  36,
  37,
  38,
  39,
  40,
  41,
]
`;

exports[`legacy > scoring 1`] = `
[
  [
    "n",
    "o",
    "t",
    "e",
  ],
  [
    "n",
    "o",
    "d",
    "e",
    "/",
    "n",
    "o",
    "t",
    "e",
  ],
  [
    0,
    1,
    7,
    8,
  ],
  true,
  [
    4,
    4,
    4,
    4,
    5,
    7,
    7,
    9,
    9,
  ],
]
`;

exports[`legacy > scoring 2`] = `
[
  0,
  1,
  7,
  8,
]
`;

exports[`legacy > scoring 3`] = `
[
  [
    "n",
    "o",
    "t",
    "e",
  ],
  [
    "n",
    "o",
    "t",
    " ",
    "o",
    "n",
    "e",
    " ",
    "t",
    "h",
    "a",
    "t",
    " ",
    "e",
    "v",
    "e",
    "n",
    "i",
    "n",
    "g",
  ],
  [
    0,
    1,
    2,
    13,
  ],
  true,
  [
    3,
    3,
    3,
    4,
    7,
    7,
    7,
    8,
    12,
    12,
    12,
    12,
    13,
    20,
    20,
    20,
    20,
    20,
    20,
    20,
  ],
]
`;

exports[`legacy > scoring 4`] = `
[
  0,
  1,
  2,
  13,
]
`;

exports[`legacy > substring not at start of word 1`] = `
[
  [
    "e",
    "r",
    ".",
    "l",
    "i",
    "f",
    "e",
    "3",
    "6",
    "0",
  ],
  [
    "d",
    "e",
    "v",
    "i",
    "c",
    "e",
    "-",
    "t",
    "r",
    "a",
    "c",
    "k",
    "e",
    "r",
    ".",
    "l",
    "i",
    "f",
    "e",
    "3",
    "6",
    "0",
    "_",
    "i",
    "p",
    "h",
    "o",
    "n",
    "e",
    "_",
    "6",
  ],
  [
    1,
    8,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
  ],
  false,
  [
    6,
    6,
    6,
    6,
    6,
    6,
    7,
    14,
    14,
    14,
    14,
    14,
    14,
    14,
    15,
    22,
    22,
    22,
    22,
    22,
    22,
    22,
    23,
    29,
    29,
    29,
    29,
    29,
    29,
    30,
    31,
  ],
]
`;

exports[`legacy > substring not at start of word 2`] = `
[
  1,
  8,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  21,
]
`;

exports[`legacy > substring not at start of word 3`] = `
[
  [
    "e",
    "r",
    ".",
    "l",
    "i",
    "f",
    "e",
    "3",
    "6",
    "0",
  ],
  [
    "s",
    "e",
    "n",
    "d",
    "o",
    "r",
    ".",
    "b",
    "a",
    "t",
    "t",
    "e",
    "r",
    "y",
    "_",
    "l",
    "i",
    "f",
    "e",
    "3",
    "6",
    "0",
    "_",
    "i",
    "p",
    "h",
    "o",
    "n",
    "e",
    "_",
    "6",
  ],
  [
    1,
    5,
    6,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
  ],
  false,
  [
    6,
    6,
    6,
    6,
    6,
    6,
    7,
    14,
    14,
    14,
    14,
    14,
    14,
    14,
    15,
    22,
    22,
    22,
    22,
    22,
    22,
    22,
    23,
    29,
    29,
    29,
    29,
    29,
    29,
    30,
    31,
  ],
]
`;

exports[`legacy > substring not at start of word 4`] = `
[
  1,
  5,
  6,
  15,
  16,
  17,
  18,
  19,
  20,
  21,
]
`;
